README
Please read the info found on the link below:
    http://blog.ufsoft.org/index.php/category/irssi-notification/

Or check the development site at:
    http://irssinotifier.ufsoft.org/


ChangeLog
Irssi Notification 0.2.0 (?/May/2007)
http://irssinotifier.ufsoft.org/svn/tags/0.2.0

14-01-2008: David Durrleman <dualmoo@gmail.com>
	* Added gentoo ebuilds
	* Removed dependency to pynotify in notifier.py

07-11-2007: David Durrleman <dualmoo@gmail.com>
	* Fixed a few typos, grammar errors, and cosmetic issues
	* Added french translation

06-11-2007: David Durrleman <dualmoo@gmail.com>
	* Improved dependency checking in setup.py

05-11-2007: Pedro Algarvio <ufs@ufsoft.org>
	* Dropped the `irclib` dependency. We now use our own unicode aware
	version of irclib.
	* We now also handle change nick for users that change their nicks into
	one we have as friend.
	* The gobject timeout is now lower, meaning, the pop-up message will be
	faster, less time will pass from the time the user send the message and we
	get the notification, this still does not mean a higher cpu usage.
	* If we pass both --gui and --debug we won't fork another process, we keep
	the same and get the logging on the console.
	* We now also make use of irclib's helper functions, so, we stop doin' the
	split stuff to know the nicks. 

18-10-2007: Pedro Algarvio <ufs@ufsoft.org>
	* Fixed broken package. Distribution egg didn't had the compiled message
	catalogs, nor, the LICENSE file.
	* Simplified and corrected the IRC codes stripping regex.

04-09-2007: Pedro Algarvio <ufs@ufsoft.org>
	* Fixed #10. Escape IRC messages with cgi.escape. Thanks vampjon.
	* Bumped RC version.

19-08-2007: Pedro Algarvio <ufs@ufsoft.org>
	* Reduced CPU usage.

24-07-2007: Pedro Algarvio <ufs@ufsoft.org>
	* Version 0.2.0RC1 Released
	* Added Babel as a dependency
	* All connections are checked if thei're connected, not just the last one
	made.
	* AWAY is now also sent to all connections, not just the last one
	connected.

----
Irssi Notification 0.1.1 (23/May/2007)
http://www.python.org/pypi/IrssiNotifier/0.1.1

24-07-2007: Pedro Algarvio <ufs@ufsoft.org>
	* Include a fully working graphic user interface.

13-07-2007: Pedro Algarvio <ufs@ufsoft.org>
	* Fixed #6, user was getting notified of all ACTION messages, not just
	those addressed to him.
	* Fully implemented David Durrleman's <dualmoo@gmail.com> patch which adds
	support for bitlbee.
	* Fixed #5, IrssiNotification is now translatable. English and Portuguese
	translations included.
	* Changed ChangeLog to be viewable as a wiki page.
	* Version Bump
	* Added info about available languages.


12-07-2007: Pedro Algarvio <ufs@ufsoft.org>
	* Added support for the user to provide a fallback charset in case he's
	not using UTF-8. (Thanks David Durrleman <dualmoo@gmail.com>)
	* Fixed #3, user no longer get's notified of his own messages if he refers
	to his own nick.
	* Added a command line argument(`--debug`) so that we get the debug messages
	from IRCLib.
	* Added Support for IRC ACTION messages.


23-05-2007: Pedro Algarvio <ufs@ufsoft.org>
	* Fixed a bug regarding private messages notification.
	Whitin a private conversation, even our own messages were beeing notified
	to us.


irssinotifier/__init__.py
# -*- coding: utf-8 -*-
# vim: sw=4 ts=4 fenc=utf-8
# =============================================================================
# $Id: __init__.py 69 2007-11-07 14:06:59Z moomoo $
# =============================================================================
#             $URL: http://irssinotifier.ufsoft.org/svn/trunk/irssinotifier/__init__.py $
# $LastChangedDate: 2007-11-07 14:06:59 +0000 (Wed, 07 Nov 2007) $
#             $Rev: 69 $
#   $LastChangedBy: moomoo $
# =============================================================================
# Copyright (C) 2007 UfSoft.org - Pedro Algarvio <ufs@ufsoft.org>
#
# Please view LICENSE for additional licensing information.
# =============================================================================

__version__ = '0.2.0'
__author__  = 'Pedro Algarvio'
__email__   = 'ufs@ufsoft.org'
__package__ = 'IrssiNotifier'
__license__ = 'BSD'
__url__     = 'http://irssinotifier.ufsoft.org'
__license_text__ = """\
Copyright (C) 2007 Unfinished Software, UfSoft.org
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in
    the documentation and/or other materials provided with the
    distribution.
 3. The name of the author may not be used to endorse or promote
    products derived from this software without specific prior
    written permission.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR "AS IS" AND ANY EXPRESS
OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
"""


irssinotifier/notifier.py
# -*- coding: utf-8 -*-
# vim: sw=4 ts=4 fenc=utf-8
# =============================================================================
# $Id: notifier.py 80 2008-01-14 03:30:32Z moomoo $
# =============================================================================
#             $URL: http://irssinotifier.ufsoft.org/svn/trunk/irssinotifier/notifier.py $
# $LastChangedDate: 2008-01-14 03:30:32 +0000 (Mon, 14 Jan 2008) $
#             $Rev: 80 $
#   $LastChangedBy: moomoo $
# =============================================================================
# Copyright (C) 2007 UfSoft.org - Pedro Algarvio <ufs@ufsoft.org>
#
# Please view LICENSE for additional licensing information.
# =============================================================================

import re
import os
from cgi import escape as cgi_escape
import sys
import xss          # This is used to know how long X has been idle
                    # http://bebop.bigasterisk.com/python
import pygtk
from irssinotifier import irclib
import threading    # This package is used to setup a timer

pygtk.require('2.0')

import dbus

if '_' not in __builtins__:
    def _(str):
        return str

IRC_CODES_RE = re.compile(
    ur'(\x02|\x06|\x07|\x11|\x12|\x13|\x0f|\x16|\x1d|\x1f|'
    ur'\x03((\d{1,2},\d{1,2})|(\d{1,2}))?)',
    re.IGNORECASE
)


class IrssiProxyNotifier:

    def __init__(self, passwd, name='', timeout=5, proxies=(), friends=(),
                 x_away=0, away_reason="", bitlbee=False, charset='latin1'):
        self.passwd = passwd
        self.name = name
        self.timeout = timeout
        self.proxies = proxies or []
        self.friends = friends or []
        self.x_away = x_away
        self.away_reason = away_reason
        self.bitlbee = bitlbee
        self.timer = None
        self.charset = charset
        irclib.FALLBACK_CHARSET = charset
        self.irc = irclib.IRC()
        self.nicks = []
        # We should only need to handle these
        self.irc.add_global_handler('privmsg', self.handle_private_messages)
        self.irc.add_global_handler('pubmsg', self.handle_public_messages)
        self.irc.add_global_handler('join', self.handle_joins)
        self.irc.add_global_handler('part', self.handle_parts)
        self.irc.add_global_handler('quit', self.handle_quits)
        self.irc.add_global_handler('nick', self.handle_nicks)
        self.irc.add_global_handler('action', self.handle_action_messages)
        self.irc.add_global_handler('mode', self.handle_modes) # for bitlbee
        #  This sets up handlers for away events, and loads the X idle tracker
        if self.x_away:
            self.idle = False
            self.away = False
            self.awaybecauseidle = False
            self.irc.add_global_handler('nowaway', self.handle_away)
            self.irc.add_global_handler('unaway', self.handle_away)
            self.tracker = xss.IdleTracker(idle_threshold=x_away)

        bus = dbus.SessionBus().get_object('org.freedesktop.Notifications',
                                           '/org/freedesktop/Notifications')
        self.bus = dbus.Interface(bus, 'org.freedesktop.Notifications')

    # TRANSLATOR: No need to translate the app name, just leave it blank
    def notify(self, message, header=_('Irssi Notifier')):
        if isinstance(message, list):
            message = u' '.join(message).strip()

        # First we try UTF-8. If the message is not valid, then we try the
        # default fallback charset.
        if not isinstance(message, unicode):
            try:
                message = unicode(message, 'utf-8')
            except UnicodeDecodeError:
                message = unicode(message, self.charset)

        uri = os.path.join(os.path.dirname(__file__), 'data', 'irssi_mini.png')
        self.bus.Notify("Irssi Notifier", 0, uri,
                        header, message.strip(), {}, {},
                        self.timeout)


    def _strip_irc_codes(self, message):
        return cgi_escape(re.sub(IRC_CODES_RE, '', message))

    def _addressing_ownnick(self, event):
        nick = irclib.nm_to_n(event.source())
        message = self._strip_irc_codes(' '.join(event.arguments()).strip())
        return nick in message

    def handle_private_messages(self, connection, event):
        nick = irclib.nm_to_n(event.source())
        if nick not in self.nicks:
            header = _("<b>New Private Message:</b>\n")
            message = self._strip_irc_codes(' '.join(event.arguments()).strip())
            message = _("<b>From <i>%(nick)s</i><tt>(%(server)s)</tt>:"
                        "</b>\n%(message)s") % \
                    {'nick': irclib.nm_to_n(event.source()),
                     'server': connection.get_server_name().rstrip('.proxy'),
                     'message': message }
            self.notify(header + message)

    def handle_public_messages(self, connection, event):
        header = _("<b>New Message:</b>\n")
        message = self._strip_irc_codes(' '.join(event.arguments()).strip())
        notify = False

        if not self._addressing_ownnick(event):
            # Are we addressed directly
            for nick in self.nicks:
                if message.startswith(nick):
                    message = message[len(nick)+1:].strip()
                    notify = True
                    # Are we instead mentioned in the message
                elif message.find(nick) != -1:
                    notify = True

        if notify:
            message = _("<b>From <i>%(nick)s</i> on <tt>%(channel)"
                        "s(%(server)s)</tt>:</b>\n%(message)s") % \
                {'nick': irclib.nm_to_n(event.source()),
                 'channel': event.target(),
                 'server': connection.get_server_name().rstrip('.proxy'),
                 'message': message }
            self.notify(header + message)

    def handle_action_messages(self, connection, event):
        header = _("<b>New IRC Action:</b>\n")
        message = self._strip_irc_codes(' '.join(event.arguments()).strip())
        notify = False
        if not self._addressing_ownnick(event):
            for nick in self.nicks:
                if nick in message:
                    notify = True

        if notify:
            message = _("<b>From <i>%(nick)s</i> on <tt>%(channel)s"
                        "(%(server)s)</tt>:</b>\n%(message)s") % \
                {'nick': irclib.nm_to_n(event.source()),
                 'channel': event.target(),
                 'server': connection.get_server_name().rstrip('.proxy'),
                 'message': '<i>%s %s</i>' % (irclib.nm_to_n(event.source()),
                                              message)}
            self.notify(header + message)


    def handle_joins(self, connection, event):
        nick = irclib.nm_to_n(event.source())
        if nick in self.friends:
            header = _('<b>Known Friend Has Joined:</b>\n')
            message = _('%(nick)s joined %(channel)s on %(server)s') % {
                'nick': nick,
                'channel': event.target(),
                'server': connection.get_server_name().rstrip('.proxy') }
            self.notify( header + message )

    def handle_parts(self, connection, event):
        nick = irclib.nm_to_n(event.source())
        if nick in self.friends:
            header = _('<b>Known Friend Has Parted:</b>\n')
            message = _('%(nick)s parted %(channel)s on %(server)s') % {
                'nick': nick,
                'channel': event.target(),
                'server': connection.get_server_name().rstrip('.proxy')}
            self.notify( header + message )

    def handle_quits(self, connection, event):
        nick = irclib.nm_to_n(event.source())
        if nick in self.friends:
            self.notify(_("'%(nick)s' has quit!") % {'nick': nick})

    def handle_nicks(self, connection, event):
        header = _("<b>Known Friend Changed Nick:</b>\n")
        newnick = event.target()
        oldnick = irclib.nm_to_n(event.source())
        notify = False
        if oldnick in self.friends:
            notify = True
            header = _("<b>Known Friend Changed Nick:</b>\n")
        elif newnick in self.friends:
            notify = True
            header = _("<b>A User Changed To A Known Friend's Nick:</b>\n")
        if notify:
            message = _("From '%(oldnick)s' to '%(newnick)s' on '%(server)s'") % {
                'oldnick': oldnick,
                'newnick': newnick,
                'server': connection.get_server_name().rstrip('.proxy')}
            self.notify( header + message )

    # This function handles modes changes to know who has come online on
    # bitlbee. A friend is online if he has mode +v on the &bitlbee channel.
    def handle_modes(self, connection, event):
        sourcenick = irclib.nm_to_n(event.source())
        if (self.bitlbee and event.target() == "&bitlbee"
        and sourcenick == "root"):
            modes = irclib.parse_channel_modes(' '.join(event.arguments()))
            for mode in modes:
                if mode[1] == "v" and mode[0] == "+":
                    header = _('<b>Contact Is Online</b>\n')
                    message = _('%(nick)s has come online on bitlbee') % {
                        'nick': mode[2] }
                    self.notify( header + message )

    # This functions handles updates the away attributes according to what the
    # server tells us.
    def handle_away(self, connection, event):
        if event.eventtype() == "nowaway":
            self.away = True
        elif event.eventtype() == "unaway":
            self.away = False
            self.awaybecauseidle = False

    # These two functions set the away status
    def setaway(self, reason):
        for connection in self.irc.connections:
            connection.send_raw(_("AWAY: %s") % reason)

    def unaway(self):
        for connection in self.irc.connections:
            connection.send_raw(_("AWAY"))

    # This functions checks if X is idle. If yes, and if the user is not
    # already away, it puts the user in away status and remembers it. If not,
    # and if the user is away and he is away because he was idle, then it puts
    # the user back from away status. It also creates a timer to check if X is
    # idle at a later time. The best timeout is given by PyXSS, so we use it.
    def check_idle(self):
        info = self.tracker.check_idle()
        if info[0] == 'idle' and not self.idle:
            if not self.away:
                self.setaway(self.away_reason)
                self.awaybecauseidle = True
            self.idle = True
        elif info[0] == 'unidle' and self.idle:
            if self.awaybecauseidle:
                self.unaway()
                self.awaybecauseidle = False
            self.idle = False
        self.timer = threading.Timer(info[1]/1000, self.check_idle)
        self.timer.start()


    def connect(self):
        for server, port, nick in self.proxies:
            self.nicks.append(nick)
            try:
                connection = self.irc.server().connect(
                    server,
                    port,
                    nick,
                    username = self.name,
                    password = self.passwd
                )
                if connection.connected:
                    self.notify(_("<b>Connection Sucessfull:</b>\n"
                                  "To irssi proxy on %(server)s:%(port)s") % {
                                    'server': server, 'port': port })
                else:
                    self.notify(_("<b>Connection Failed:</b>\nFailed to "
                                  "connect to %(server)s:%(port)s") % {
                                    'server': server, 'port': port })
            except irclib.ServerConnectionError, err:
                print err
                sys.exit(1)

    def start(self):
        if self.x_away:
            self.check_idle()

    def is_connected(self):
        connected = False
        for connection in self.irc.connections:
            #print connection, connection.__dict__
            if connection.connected:
                connected = True
        return connected

    def process(self, timeout=0):
        if self.is_connected():
            self.irc.process_once(timeout)
            return True
        return False

    def process_non_gui(self):
        while self.is_connected():
            #print "processing non GUI"
            self.irc.process_once(0.2)
        self.notify(_("Disconnected from server"))
        print _("Disconnected from server")
        self.quit()
        sys.exit(1)

    def quit(self):
        self.irc.disconnect_all()
        if self.timer:
            self.timer.cancel() # We kill the timer before exiting.


irssinotifier/translation.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# vim: sw=4 ts=4 fenc=utf-8
# =============================================================================
# $Id: translation.py 19 2007-07-24 18:16:48Z s0undt3ch $
# =============================================================================
#             $URL: http://irssinotifier.ufsoft.org/svn/trunk/irssinotifier/translation.py $
# $LastChangedDate: 2007-07-24 19:16:48 +0100 (Tue, 24 Jul 2007) $
#             $Rev: 19 $
#   $LastChangedBy: s0undt3ch $
# =============================================================================
# Copyright (C) 2007 UfSoft.org - Pedro Algarvio <ufs@ufsoft.org>
#
# Please view LICENSE for additional licensing information.
# =============================================================================

import os
import gettext
import locale
import __builtin__

def set_lang(lang, codeset='utf-8', gui=False):
    if not isinstance(lang, list):
        lang = [lang]

    locale_dir = os.path.join(os.path.dirname(__file__), 'i18n')
    domain = 'irssinotifier'

    gettext.bindtextdomain(domain, locale_dir)
    gettext.textdomain(domain)
    locale.bindtextdomain(domain, locale_dir)
    locale.textdomain(domain)
    if gui:
        import gtk.glade
        gtk.glade.bindtextdomain(domain)
        gtk.glade.textdomain(domain)

    try:
        translator = gettext.translation(domain,
                                         locale_dir,
                                         languages=lang,
                                         codeset=codeset)

    except IOError, ioe:
        lang = ['en']
        print 'Language not supportted: %r' % ioe
        print 'Fallback to english'
        translator = gettext.translation(domain,
                                         locale_dir,
                                         languages=lang,
                                         codeset=codeset)

    if lang[0] != 'en' and gui:
        # We only need to do this because the Glade GUI reads info
        # from the environment
        new_locale = '.'.join([lang[0], locale.getdefaultlocale()[1]])
        try:
            os.environ['LC_ALL'] = new_locale
            locale.setlocale(locale.LC_ALL,'')
        except locale.Error:
            import sys
            print "Locale %r is not supported" % new_locale
            print "Please update/generate your system locales with the new"
            print "locale you want to use"
            sys.exit(1)

    __builtin__._ = translator.ugettext
    __builtin__.__glade_translator__ = locale.textdomain(domain)
    # Late import to make use of locale set
    from irssinotifier import notifier, ui



irssinotifier/parser.py
# -*- coding: utf-8 -*-
# vim: sw=4 ts=4 fenc=utf-8
# =============================================================================
# $Id: parser.py 72 2007-11-07 17:25:16Z s0undt3ch $
# =============================================================================
#             $URL: http://irssinotifier.ufsoft.org/svn/trunk/irssinotifier/parser.py $
# $LastChangedDate: 2007-11-07 17:25:16 +0000 (Wed, 07 Nov 2007) $
#             $Rev: 72 $
#   $LastChangedBy: s0undt3ch $
# =============================================================================
# Copyright (C) 2007 UfSoft.org - Pedro Algarvio <ufs@ufsoft.org>
#
# Please view LICENSE for additional licensing information.
# =============================================================================

import os
import sys
from optparse import OptionParser
import ConfigParser
from irssinotifier import __version__ as VERSION

class IrssiProxyNotifierStartup:
    def __init__(self):
        self.cfgfile = None
        parser = OptionParser(
            usage = "%prog [options]",
            version = "%prog " + VERSION
        )
        parser.description = "%prog is intended to give you a visual, non " + \
        "obtrusive notification of messages addressed to you or mentioned " + \
        "to you"
        parser.add_option(
            '--config', '-C',
            dest = 'configfile',
            metavar = 'PATH_TO_FILE',
            help = "Provide an alternate configuration file. "
            "Default: '~/.irssinotifier'"
        )
        parser.add_option(
            '--nick', '-n',
            dest = 'nick',
            default = '',
            help = "Global nickname to pass to irssi-proxy. Use it if your "
            "nick is the same for all networks. Default: '%default'"
        )
        parser.add_option(
            '--passwd', '-p',
            dest = 'passwd',
            help = "Irssi proxy password"
        )
        parser.add_option(
            '--name', '-N',
            dest = 'name',
            default = '',
            help = "Your Name. Default: '%default'"
        )
        parser.add_option(
            '--proxies', '-P',
            action = 'append',
            dest = 'proxies',
            default = [],
            metavar = 'ADDRESS:PORT[:NICK]',
            help = "Pass an irssi-proxy address:port to be notified of. "
            "Repeat the argument to add more servers. You can optionally "
            "pass the nick for that network if your not using the global "
            "one and your nicks differ from network to netowrk. Default: %default"
        )
        parser.add_option(
            '--friend', '-F',
            action = 'append',
            dest = 'friends',
            metavar = 'FRIEND',
            default = [],
            help = "List of friends to be notified of parts, joins and quits. "
            "One friend per argument call. Default: %default"
        )
        parser.add_option(
            '--timeout', '-T',
            type = 'int',
            dest = 'timeout',
            metavar = 'SECONDS',
            default = 5,
            help = "Notification pop-up timeout (in seconds). Default: %default"
        )
        # These options allow the user to be automatically put in away mode
        # when X has been idle for some time.
        parser.add_option(
            '--x-away', '-X',
            type = int,
            dest = 'x_away',
            metavar = 'SECONDS',
            default = 0,
            help = "Enable auto-away according to X idle time (0 to disable). "
            "Default: %default"
        )
        parser.add_option(
            '--x-away-reason', '-R',
            dest = 'away_reason',
            default = 'inactivity',
            help = "The away reason to use in case of X inactivity. Default:"
            "'%default' "
        )
        # If this option is set, the server is known to be of bitlbee type,
        # and we will check &bitlbee to see if friends come online.
        parser.add_option(
            '--bitlbee', '-b',
            action = 'store_true',
            default = False,
            dest = 'bitlbee',
            help = "Enable notifications of away/joins/quits on a "
                   "bitlbee server"
        )
        # This option is the charset to fallback to when text received is not
        # in UTF-8.
        parser.add_option(
            '--fallback-charset', '-c',
            dest = 'charset',
            default = None,
            help = "The charset to fallback to when messages received are not"
                   "in UTF-8. Default: '%default' "
        )
        parser.add_option(
            '--language', '-l',
            dest='language',
            help="Use the specified language translation. "
                 "Available Languages: 'en', 'fr_FR', 'pt_PT'"
        )
        parser.add_option(
            '--write-configs', '-W',
            action = 'store_true',
            default = False,
            dest = 'write_configs',
            help = "Write the options to the configuration file "
            "'~/.irssinotifier'"
        )
        parser.add_option(
            '--debug',
            action='store_true',
            default=False,
            dest='debug',
            help='Output IRCLib debug messages(extremely verbose)')
        parser.add_option(
            '--gui',
            action = 'store_true',
            default = False,
            dest = 'gui',
            help = 'run graphical user interface')
        self.parser = parser

    def parse_args(self, configfile='~/.irssinotifier'):
        """First parse the user's configuration file for options, if available
        make them the defaults for the parser
        """
        cfgfile = os.path.expanduser(configfile)
        if os.path.exists(cfgfile):
            config = ConfigParser.SafeConfigParser()
            config.read(cfgfile)
            if config.has_section('main'):
                opts = {}
                for opt in (
                    'nick', 'passwd', 'name', 'proxies', 'friends', 'timeout',
                    'x_away', 'away_reason', 'charset', 'bitlbee', 'gui',
                    'language', 'debug'):
                    try:
                        if opt in ('proxies', 'friends'):
                            opts[opt] = config.get('main', opt).split()
                        elif opt in ('timeout', 'x_away'):
                            try:
                                opts[opt] = config.getint('main', opt)
                            except ValueError:
                                opts[opt] = int(config.getfloat('main',
                                                                'timeout'))
                        elif opt in ('bitlbee', 'gui', 'debug'):
                            opts[opt] = config.getboolean('main', opt)
                        else:
                            opts[opt] = config.get('main', opt)
                    except ConfigParser.NoOptionError:
                        continue
                if opts:
                    self.parser.set_defaults(**opts)
        self.cfgfile = cfgfile
        return self.parser.parse_args()

    def run(self):
        (options, _) = self.parse_args()
        if options.configfile:
            # We must re-parse
            (options, _) = self.parse_args(configfile=options.configfile)
        else:
            options.configfile = '~/.irssinotifier'

        if options.debug:
            from irssinotifier import irclib
            irclib.DEBUG = True

        if options.charset:
            charset = options.charset
        else:
            charset = 'utf-8'

        proxies = []
        for proxy in options.proxies:
            t = proxy.split(':')
            if len(t) == 3:
                # Also passed User
                entry = (t[0], int(t[1]), t[2])
                if entry not in proxies:
                    proxies.append(entry)
            elif len(t) == 2:
                # No user passed
                if not options.nick:
                    self.parser.error("You must provide your nick."
                        "Either pass it on --proxies or on --nick"
                    )
                entry = (t[0], int(t[1]), options.nick)
                if entry not in proxies:
                    proxies.append(entry)
            else:
                self.parser.error("Bad arguments passed to --proxies %r" % proxy)
        friends = []
        for friend in options.friends:
            if friend not in friends:
                friends.append(friend)

        timeout = options.timeout * 1000
        # We need an ms value.
        x_away = options.x_away * 1000

        if options.write_configs:
            print "Writing configuration to %r ..." % options.configfile
            cfgfile = os.path.expanduser(options.configfile)
            cfgfile = open(cfgfile, 'w')
            config = ConfigParser.SafeConfigParser()
            config.add_section('main')
            print 'passwd\t...',
            if options.passwd:
                print 'OK'
                config.set('main', 'passwd', options.passwd)
            else:
                print 'no %r option set!' % 'passwd'
            print 'proxies\t...',
            local_proxies = []
            for host, port, user in proxies:
                entry = '%s:%s:%s' % ( host, port, user )
                if entry not in local_proxies:
                    local_proxies.append(entry)
            if local_proxies:
                config.set('main', 'proxies', ' '.join(local_proxies))
                print 'OK'
            else:
                print 'no %r option set!' % 'proxies'
            print 'timeout\t...',
            if options.timeout:
                config.set('main', 'timeout', str(options.timeout))
                print 'OK'
            else:
                print 'no %r option set!' % 'timeout'
            print 'x-away\t...',
            if options.x_away:
                config.set('main', 'x_away', str(options.x_away))
                print 'OK'
            else:
                print 'no %r option set!' % 'x-away'
            print 'reason\t...',
            if options.away_reason:
                config.set('main', 'away_reason', options.away_reason)
                print 'OK'
            else:
                print 'no %r option set!' % 'x-away-reason'
            if options.charset:
                config.set('main', 'charset', options.charset)
                print 'OK'
            else:
                print 'no %r option set!' % 'fallback-charset'
            print 'friends\t...',
            if friends:
                config.set('main', 'friends', ' '.join(friends))
                print 'OK'
            else:
                print 'no %r option set!' % 'friends'
            print 'bitlbee\t...',
            if options.bitlbee:
                config.set('main', 'bitlbee', str(options.bitlbee))
                print 'OK'
            else:
                print 'no %r option set!' % 'bitlbee'
            config.write(cfgfile)
            print "Done!"
            sys.exit(1)

        if not options.passwd:
            self.parser.error("You must provide the irssi-proxy password")

        if not options.proxies:
            self.parser.error(
                "You must pass at least one irssi-proxy addr:port"
            )

        if not options.language:
            from locale import getdefaultlocale
            language = getdefaultlocale()[0]
            if language in ('en', 'en_GB', 'en_US'):
                language = 'en'
            if language in ('en', 'fr_FR', 'pt_PT'):
                options.language = language
        elif options.language:
            if not options.language in ('en', 'fr_FR', 'pt_PT'):
                self.parser.error('Language not supported')

        from irssinotifier.translation import set_lang
        set_lang(options.language, charset, options.gui)
        # Late import so translations are not ignored
        from irssinotifier.notifier import IrssiProxyNotifier

        notifier = IrssiProxyNotifier(
            options.passwd,
            name=options.name,
            timeout=timeout,
            proxies=proxies,
            friends=friends,
            charset=options.charset,
            x_away=x_away,
            away_reason=options.away_reason,
            bitlbee=options.bitlbee
        )
        notifier.connect()
        if options.gui:
            from irssinotifier.ui import TrayApp
            ta = TrayApp(options=options, cfgfile=self.cfgfile,
                         notifier=notifier)

            if options.debug:
                try:
                    ta.main()
                finally:
                    ta.exit_app()
                    sys.exit(0)
            else:
                # Fork a new process for the GUI
                # http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/66012
                try:
                    pid = os.fork()
                    if pid > 0:
                        # exit first parent
                        sys.exit(0)
                except OSError, e:
                    print >>sys.stderr, "fork #1 failed: %d (%s)" % (e.errno, e.strerror)
                    sys.exit(1)

                # decouple from parent environment
                os.chdir("/")
                os.setsid()
                os.umask(0)

                # do second fork
                try:
                    pid = os.fork()
                    if pid > 0:
                        # exit from second parent, print eventual PID before
                        print "Daemon PID %d" % pid
                        sys.exit(0)
                except OSError, e:
                    print >>sys.stderr, "fork #2 failed: %d (%s)" % (e.errno, e.strerror)
                    sys.exit(1)

                ta.main()
        else:
            if options.debug:
                try:
                    notifier.start()
                    notifier.process_non_gui()
                except KeyboardInterrupt:
                    notifier.quit()
                    sys.exit(1)
            else:
                try:
                    notifier.start()
                    notifier.process_non_gui()
                finally:
                    notifier.quit()
                    sys.exit(1)

def main():
    ipn = IrssiProxyNotifierStartup()
    ipn.run()
